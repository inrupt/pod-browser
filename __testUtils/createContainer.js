/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  createSolidDataset,
  createThing,
  addDatetime,
  addDecimal,
  addInteger,
  setThing,
} from "@inrupt/solid-client";
import { rdf, ldp } from "rdf-namespaces";

export const TIMESTAMP = new Date(Date.UTC(2020, 5, 2, 15, 59, 21));

export default function createContainer(
  url = "https://user.dev.inrupt.net/",
  type = ldp.BasicContainer
) {
  const thing = createThing({ url });
  const publicContainer = [
    (t) => addUrl(t, rdf.type, type),
    (t) => addUrl(t, rdf.type, ldp.Container),
    (t) => addDatetime(t, "http://purl.org/dc/terms/modified", TIMESTAMP),
    (t) =>
      addDecimal(t, "http://www.w3.org/ns/posix/stat#mtime", 1591131561.195),
    (t) => addInteger(t, "http://www.w3.org/ns/posix/stat#size", 4096),
    (t) => addUrl(t, ldp.contains, "https://user.dev.inrupt.net/public/games/"),
  ].reduce((acc, fn) => fn(acc), thing);

  return setThing(createSolidDataset(), publicContainer);
}
