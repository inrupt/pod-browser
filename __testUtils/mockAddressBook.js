/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createSolidDataset,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import { acl, dc, rdf } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { aliceWebIdUrl } from "./mockPersonResource";
import { vcardExtras } from "../src/addressBook";
import { getAddressBookIndexDefaultUrl } from "../src/models/addressBook";
import { joinPath } from "../src/stringHelpers";

export const defaultContainerUrl = "https://user.example.com/contacts";
export const defaultAddressBookDatasetUrl = joinPath(
  defaultContainerUrl,
  "index.ttl"
);
export const defaultAddressBookThingUrl = `${defaultAddressBookDatasetUrl}#this`;

export function mockAddressBookThing(options = {}) {
  const containerUrl = options.containerUrl || defaultContainerUrl;
  const owner = options.owner || aliceWebIdUrl;
  const indexUrl =
    options.indexUrl || getAddressBookIndexDefaultUrl(containerUrl);
  const mainIndexUrl = `${indexUrl}#this`;
  return chain(
    mockThingFrom(mainIndexUrl),
    (t) => addUrl(t, rdf.type, vcardExtras("AddressBook")),
    (t) => addUrl(t, acl.owner, owner),
    (t) => addStringNoLocale(t, dc.title, "Contacts")
  );
}

export function mockAddressBookDataset(
  options = {},
  thing = mockAddressBookThing(options)
) {
  const containerUrl = options.containerUrl || defaultContainerUrl;
  const indexUrl =
    options.indexUrl || getAddressBookIndexDefaultUrl(containerUrl);
  return chain(mockSolidDatasetFrom(indexUrl), (d) => setThing(d, thing));
}

export default function mockAddressBook(options = {}) {
  const containerUrl = options.containerUrl || defaultContainerUrl;
  const thing = mockAddressBookThing(options);
  return {
    containerUrl,
    thing,
    dataset: mockAddressBookDataset(options, thing),
  };
}

export function mockEmptyAddressBook(containerUrl, owner, title = "Contacts") {
  const thing = chain(
    mockAddressBookThing({ containerUrl, owner }),
    (t) => addUrl(t, rdf.type, vcardExtras("AddressBook")),
    (t) => addUrl(t, acl.owner, owner),
    (t) => addStringNoLocale(t, dc.title, title)
  );
  return {
    containerUrl,
    dataset: setThing(createSolidDataset(), thing),
    thing,
  };
}
