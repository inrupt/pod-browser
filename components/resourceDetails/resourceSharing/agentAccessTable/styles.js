/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { createStyles } from "@solid/lit-prism-patterns";

export default function styles(theme) {
  return createStyles(theme, ["icons", "table"], {
    accordion: {
      padding: theme.spacing(1.8, 1.6),
      color: theme.palette.secondary.contrastText,
    },
    rounded: {
      borderRadius: theme.shape.borderRadius,
      boxShadow: `0 1px 4px 0 ${theme.palette.grey[500]}`,
    },
    title: {
      fontSize: theme.typography.h4.fontSize,
      fontWeight: 800,
    },
    headerContainer: {
      display: "flex",
      flexDirection: "row",
      alignItems: "center",
      marginBottom: theme.spacing(1.6),
      "&& p": {
        padding: 0,
        margin: 0,
      },
    },
    textContainer: {
      width: "100%",
      flexDirection: "column",
    },
    description: {
      fontSize: "0.8125rem",
    },
    permissionsContainer: {
      position: "relative",
      boxSizing: "border-box",
      border: `1px solid ${theme.palette.grey.A100}`,
      borderRadius: theme.shape.borderRadius,
      fontSize: theme.typography.body2.fontSize,
      padding: 0,
      overflow: "auto",
      "&& p": {
        padding: 0,
        margin: 0,
      },
    },
    showAllButtonContainer: {
      width: "100%",
      background:
        "linear-gradient(180deg, rgba(255,255,255,0) 0%, #FFFFFF 100%)",
      position: "absolute",
      textAlign: "center",
      bottom: 0,
      padding: theme.spacing(0.6, 0.7),
    },
    showAllButton: {
      cursor: "pointer",
      width: "100%",
      backgroundColor: "transparent",
      border: "none",
    },
    expanded: {
      position: "relative",
    },
    showAllText: {
      textAlign: "center",
      color: theme.palette.primary.main,
      fontSize: "13px",
    },
    showAllButtonIcon: {
      color: theme.palette.primary.main,
      fontSize: "13px",
    },
    titleAndButtonContainer: {
      display: "flex",
      flexDirection: "row",
      justifyContent: "space-between",
    },
    icon: {
      padding: theme.spacing(1),
      borderRadius: "50%",
      marginRight: theme.spacing(1.6),
      fontSize: theme.typography.h1.fontSize,
    },
    iconViewer: {
      color: "#2196F3",
      backgroundColor: "#E8F4FD",
    },
    iconBlocked: {
      color: "#F44336",
      backgroundColor: "#FDECEA",
    },
    iconEditor: {
      color: "#4CAF50",
      backgroundColor: "#EDF7ED",
    },
    emptyStateTextContainer: {
      display: "flex",
      flexDirection: "column",
      padding: theme.spacing(0.7, 1.4),
      textAlign: "center",
    },
    "agents-table": {
      padding: 0,
    },
    "agent-cell": {
      whiteSpace: "break-spaces",
      padding: 0,
    },
    tabsContainer: {
      paddingLeft: theme.spacing(1.2),
      paddingRight: theme.spacing(1.2),
      borderBottom: `1px solid ${theme.palette.grey.A100}`,
    },
    tab: {
      textTransform: "none",
      minWidth: "max-content",
      fontFamily: theme.typography.body1.fontFamily,
      fontSize: theme.typography.body1.fontSize,
      color: theme.palette.primary.text,
    },
    selected: {
      color: theme.palette.primary.main,
    },
    indicator: {
      height: "3px",
      width: "100%",
      backgroundColor: theme.palette.primary.main,
    },
    searchBoxContainer: {
      display: "flex",
      margin: theme.spacing(0.7, 1.2),
      border: "1px solid #D8D8D8",
      borderRadius: "10px",
      height: "2.5rem",
    },
    searchInput: {
      fontSize: "0.8125rem",
      width: "100%",
      font: "inherit",
      fontWeight: 500,
      color: theme.palette.primary.text,
    },
    iconSearch: {
      fontSize: theme.typography.body1.fontSize,
    },
  });
}
