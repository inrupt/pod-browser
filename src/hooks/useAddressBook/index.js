/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { useContext, useEffect, useState } from "react";
import { fetchLitDataset } from "@inrupt/solid-client";
import SessionContext from "../../contexts/sessionContext";
import useAuthenticatedProfile from "../useAuthenticatedProfile";
import { contactsContainerIri, saveNewAddressBook } from "../../addressBook";
import { getResource } from "../../solidClientHelpers/resource";
import { ERROR_CODES, isHTTPError } from "../../solidClientHelpers/utils";

export default function useAddressBook() {
  const [addressBook, setAddressBook] = useState(null);
  const [error, setError] = useState(null);
  const { session } = useContext(SessionContext);
  const profile = useAuthenticatedProfile();

  useEffect(() => {
    if (!session.info.isLoggedIn || !profile) return;
    const { pods, webId } = profile;
    const { fetch } = session;
    const contactsIri = contactsContainerIri(pods[0]);

    (async () => {
      let addressBookUri;

      const {
        response: existingAddressBook,
        error: existingError,
      } = await getResource(contactsIri, fetch);

      if (existingAddressBook) {
        setAddressBook(existingAddressBook.dataset);
        return;
      }

      if (existingError && !isHTTPError(existingError, ERROR_CODES.NOT_FOUND)) {
        const {
          response: newAddressBook,
          error: newError,
        } = await saveNewAddressBook(
          {
            iri: contactsIri,
            owner: webId,
          },
          fetch
        );
        if (newError) {
          setError(newError);
          return;
        }
        addressBookUri = newAddressBook.iri;
      } else if (existingError) {
        setError(existingError);
        return;
      }

      setAddressBook(await fetchLitDataset(addressBookUri, { fetch }));
    })();
  }, [session, profile]);

  return [addressBook, error];
}
