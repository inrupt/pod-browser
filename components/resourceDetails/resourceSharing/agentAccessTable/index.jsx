/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable react/jsx-one-expression-per-line */

import React, { useMemo, useState } from "react";
import PropTypes from "prop-types";
import { useFilters, useGlobalFilter, useTable } from "react-table";
import { useBem } from "@solid/lit-prism-patterns";
import clsx from "clsx";
import { Accordion, createStyles, Typography } from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import useNamedPolicyPermissions from "../../../../src/hooks/useNamedPolicyPermissions";
import usePermissionsWithProfiles from "../../../../src/hooks/usePermissionsWithProfiles";
import AgentAccess from "../agentAccess";
import AddAgentButton from "../addAgentButton";
import AgentsTableTabs from "../agentsTableTabs";
import AgentsSearchBar from "../agentsSearchBar";

import styles from "./styles";
import PolicyHeader from "../policyHeader";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));
const TESTCAFE_ID_SHOW_ALL_BUTTON = "show-all-button";
const TESTCAFE_ID_HIDE_BUTTON = "hide-button";
export const TESTCAFE_ID_AGENT_ACCESS_TABLE = "agent-access-table";

export default function AgentAccessTable({ type }) {
  const {
    data: namedPermissions,
    mutate: mutatePermissions,
  } = useNamedPolicyPermissions(type);

  const { permissionsWithProfiles: permissions } = usePermissionsWithProfiles(
    namedPermissions
  );

  const editorsDescription = (
    <p>
      <b>Can </b>
      view, edit and delete this resource
    </p>
  );

  const viewersDescription = (
    <p>
      <b>Can </b>
      view but
      <b> cannot </b>
      edit or delete this resource
    </p>
  );

  const blockedDescription = (
    <p>
      <b>Cannot </b>
      view this resource
    </p>
  );

  const PERMISSIONS_TYPE_MAP = {
    editors: {
      icon: "icon-editor",
      iconClassName: "iconEditor",
      title: "Editors",
      emptyStateText: "No editors",
      description: editorsDescription,
    },
    viewers: {
      icon: "icon-view",
      iconClassName: "iconViewer",
      title: "Viewers",
      emptyStateText: "No viewers",
      description: viewersDescription,
    },
    blocked: {
      icon: "icon-block",
      iconClassName: "iconBlocked",
      title: "Blocked",
      emptyStateText: "No one is blocked",
      description: blockedDescription,
    },
  };

  const bem = useBem(useStyles());

  const classes = useStyles();

  const [showAll, setShowAll] = useState(false);
  const [selectedTabValue, setSelectedTabValue] = useState("");

  const { emptyStateText } = PERMISSIONS_TYPE_MAP[type];

  const columns = useMemo(
    () => [
      {
        header: "",
        accessor: "profile.name",
        disableSortBy: true,
        modifiers: ["align-center", "width-preview"],
      },
      {
        header: "",
        accessor: "webId",
        disableSortBy: true,
        modifiers: ["align-center", "width-preview"],
      },
      {
        header: "",
        accessor: "profile.types",
        disableSortBy: true,
        modifiers: ["align-center", "width-preview"],
      },
    ],
    []
  );

  const data = useMemo(() => {
    if (!permissions) {
      return [];
    }

    return showAll ? permissions : permissions.slice(0, 3);
  }, [permissions, showAll]);

  const {
    getTableProps,
    getTableBodyProps,
    rows,
    prepareRow,
    setFilter,
    setGlobalFilter,
  } = useTable(
    {
      columns,
      data,
      defaultCanSort: true,
    },
    useFilters,
    useGlobalFilter
  );

  const handleFilterChange = (e) => {
    const { value } = e.target;
    setGlobalFilter(value || undefined);
  };

  /* istanbul ignore next */
  const handleTabChange = (e, newValue) => {
    setSelectedTabValue(newValue);
    // TODO: this filter will change once we have groups - ignoring from test coverage until it is functional
    setFilter("profile.types", newValue);
  };

  return (
    <Accordion
      defaultExpanded
      classes={{ root: classes.accordion, rounded: classes.rounded }}
      data-testid={TESTCAFE_ID_AGENT_ACCESS_TABLE}
    >
      <PolicyHeader type={type} isPolicyList>
        <AddAgentButton
          type={type}
          mutatePermissions={mutatePermissions}
          permissions={permissions}
        />
      </PolicyHeader>
      <div className={classes.permissionsContainer}>
        {!!permissions.length && (
          <>
            <AgentsTableTabs
              handleTabChange={handleTabChange}
              selectedTabValue={selectedTabValue}
            />
            <AgentsSearchBar handleFilterChange={handleFilterChange} />
          </>
        )}
        {permissions.length ? (
          <table
            className={clsx(bem("table"), bem("agents-table"))}
            {...getTableProps()}
          >
            <tbody className={bem("table__body")} {...getTableBodyProps()}>
              {rows.length ? (
                rows.map((row) => {
                  prepareRow(row);
                  const details = row.original;
                  return (
                    <tr className={bem("table__body-row")} key={details.webId}>
                      <td
                        className={clsx(
                          bem("table__body-cell"),
                          bem("agent-cell")
                        )}
                      >
                        <AgentAccess
                          permission={details}
                          mutatePermissions={mutatePermissions}
                        />
                      </td>
                    </tr>
                  );
                })
              ) : (
                <span className={classes.emptyStateTextContainer}>
                  <p>
                    {selectedTabValue === ""
                      ? emptyStateText
                      : `No ${
                          selectedTabValue === "Person" ? "people " : "groups "
                        } found`}
                  </p>
                </span>
              )}
            </tbody>
          </table>
        ) : (
          <span className={classes.emptyStateTextContainer}>
            <p>{emptyStateText}</p>
          </span>
        )}
        {permissions.length > 3 && (
          <div
            className={clsx(
              classes.showAllButtonContainer,
              showAll && bem("expanded")
            )}
          >
            <button
              data-testid={
                showAll ? TESTCAFE_ID_HIDE_BUTTON : TESTCAFE_ID_SHOW_ALL_BUTTON
              }
              type="button"
              className={classes.showAllButton}
              onClick={() => setShowAll(!showAll)}
            >
              <Typography classes={{ body1: classes.showAllText }}>
                {showAll ? (
                  <span>
                    Hide{" "}
                    <i
                      className={clsx(
                        bem("icon-caret-up"),
                        classes.showAllButtonIcon
                      )}
                    />
                  </span>
                ) : (
                  <span>
                    Show all ({permissions.length}){" "}
                    <i
                      className={clsx(
                        bem("icon-caret-down"),
                        classes.showAllButtonIcon
                      )}
                    />
                  </span>
                )}
              </Typography>
            </button>
          </div>
        )}
      </div>
    </Accordion>
  );
}

AgentAccessTable.propTypes = {
  type: PropTypes.string.isRequired,
};
