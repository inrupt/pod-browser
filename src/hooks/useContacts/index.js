/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { useSession } from "@inrupt/solid-ui-react";
import useSWR from "swr";
import { thingAsMarkdown } from "@inrupt/solid-client";
import useAddressBook from "../useAddressBook";
import { getContactAll } from "../../models/contact";
import { compareArray } from "../../models/array";

function reduceToString(array) {
  return array.reduce(
    (memo, item) => memo.concat(thingAsMarkdown(item.thing)),
    ""
  );
}

export default function useContacts(type) {
  const { fetch } = useSession();
  const { data: addressBook, error: addressBookError } = useAddressBook();
  return useSWR(
    ["contacts", addressBook, type],
    async () => {
      if (!addressBook && !addressBookError) return null;
      if (addressBookError) throw addressBookError;
      return getContactAll(addressBook, [type], fetch);
    },
    {
      // revalidateOnMount: true,
      // revalidateOnFocus: false,
      // revalidateOnReconnect: false,
      // refreshInterval: 0,
      shouldRetryOnError: false,
      // errorRetryCount: 0,
      compare: (a, b) =>
        compareArray(a, b, (c, d) => reduceToString(c) === reduceToString(d)),
      // compare: (a, b) => {
      //   if (typeof a === "undefined") {
      //     return typeof b === "undefined";
      //   }
      //   if (typeof b === "undefined") {
      //     return typeof a === "undefined";
      //   }
      //   if (!Array.isArray(a) || !Array.isArray(b)) {
      //     return false;
      //   }
      //   if (a.length !== b.length) {
      //     return false;
      //   }
      //   const reduce = reduceToString(a) === reduceToString(b);
      //   // if (!reduce) {
      //   //   console.log("COMPARE ARRAY A", reduceToString(a));
      //   //   console.log("COMPARE ARRAY B", reduceToString(b));
      //   // }
      //   return reduce;
      // },
    }
  );
}
