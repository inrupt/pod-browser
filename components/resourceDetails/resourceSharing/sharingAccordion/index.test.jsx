/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { renderWithTheme } from "../../../../__testUtils/withTheme";
import mockPersonContact from "../../../../__testUtils/mockPersonContact";
import mockAddressBook from "../../../../__testUtils/mockAddressBook";
import SharingAccordion from "./index";
import { TESTCAFE_ID_AGENT_ACCESS_TABLE } from "../agentAccessTable";
import { TESTCAFE_ID_ADVANCED_SHARING_BUTTON } from "../advancedSharingButton";
import useContacts from "../../../../src/hooks/useContacts";
import useAddressBook from "../../../../src/hooks/useAddressBook";

jest.mock("../../../../src/hooks/useContacts");
const mockedUseContacts = useContacts;

jest.mock("../../../../src/hooks/useAddressBook");
const mockedUseAddressBook = useAddressBook;

describe("SharingAccordion", () => {
  beforeEach(() => {
    mockedUseAddressBook.mockReturnValue({ data: mockAddressBook() });
    mockedUseContacts.mockReturnValue({
      data: [
        mockPersonContact(
          mockAddressBook(),
          "https://example.org/contacts/Person/1234/",
          "Example 1"
        ),
        mockPersonContact(
          mockAddressBook(),
          "https://example.org/contacts/Person/3456/",
          "Example 2"
        ),
      ],
    });
  });
  // Note: since the permissions cannot be mocked reliably for the custom policies, those tested separately in the table component
  test("it renders three lists of named policies for editors, viewers and blocked and an Advanced Sharing button", () => {
    const { asFragment, queryAllByTestId, queryByTestId } = renderWithTheme(
      <SharingAccordion />
    );
    expect(queryAllByTestId(TESTCAFE_ID_AGENT_ACCESS_TABLE)).toHaveLength(3);
    expect(queryByTestId(TESTCAFE_ID_ADVANCED_SHARING_BUTTON)).not.toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });
});
