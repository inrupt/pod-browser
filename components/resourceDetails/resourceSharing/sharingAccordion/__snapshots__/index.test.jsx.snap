// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgentAccessList it renders three lists of empty permissions for editors, viewers and blocked 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-root PodBrowser-root PodBrowser-accordion PodBrowser-expanded PodBrowser-rounded PodBrowser-rounded PodBrowser-elevation1 PodBrowser-rounded"
  >
    <div
      class="PodBrowser-headerContainer"
    >
      <i
        class="PodBrowser-icon-editor PodBrowser-icon PodBrowser-iconEditor"
      />
      <div
        class="PodBrowser-textContainer"
      >
        <div
          class="PodBrowser-titleAndButtonContainer"
        >
          <p
            class="PodBrowser-title"
          >
            Editors
          </p>
          <button
            class="PodBrowser-button PodBrowser-button--action PodBrowser-button"
            data-testid="add-agent-button"
          >
            <i
              alt="Add Editors Button"
              class="PodBrowser-icon-add PodBrowser-icon"
            />
            Add Editors
          </button>
        </div>
        <span
          class="PodBrowser-description"
        >
          <p>
            <b>
              Can 
            </b>
            view, edit and delete this resource
          </p>
        </span>
      </div>
    </div>
    <div
      class="PodBrowser-container PodBrowser-entered"
      style="min-height: 0px;"
    >
      <div
        class="PodBrowser-wrapper"
      >
        <div
          class="PodBrowser-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="PodBrowser-permissionsContainer"
            >
              <span
                class="PodBrowser-emptyStateTextContainer"
              >
                <p>
                  No editors
                </p>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="PodBrowser-root PodBrowser-root PodBrowser-accordion PodBrowser-expanded PodBrowser-rounded PodBrowser-rounded PodBrowser-elevation1 PodBrowser-rounded"
  >
    <div
      class="PodBrowser-headerContainer"
    >
      <i
        class="PodBrowser-icon-view PodBrowser-icon PodBrowser-iconViewer"
      />
      <div
        class="PodBrowser-textContainer"
      >
        <div
          class="PodBrowser-titleAndButtonContainer"
        >
          <p
            class="PodBrowser-title"
          >
            Viewers
          </p>
          <button
            class="PodBrowser-button PodBrowser-button--action PodBrowser-button"
            data-testid="add-agent-button"
          >
            <i
              alt="Add Viewers Button"
              class="PodBrowser-icon-add PodBrowser-icon"
            />
            Add Viewers
          </button>
        </div>
        <span
          class="PodBrowser-description"
        >
          <p>
            <b>
              Can 
            </b>
            view but
            <b>
               cannot 
            </b>
            edit or delete this resource
          </p>
        </span>
      </div>
    </div>
    <div
      class="PodBrowser-container PodBrowser-entered"
      style="min-height: 0px;"
    >
      <div
        class="PodBrowser-wrapper"
      >
        <div
          class="PodBrowser-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="PodBrowser-permissionsContainer"
            >
              <span
                class="PodBrowser-emptyStateTextContainer"
              >
                <p>
                  No viewers
                </p>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="PodBrowser-root PodBrowser-root PodBrowser-accordion PodBrowser-expanded PodBrowser-rounded PodBrowser-rounded PodBrowser-elevation1 PodBrowser-rounded"
  >
    <div
      class="PodBrowser-headerContainer"
    >
      <i
        class="PodBrowser-icon-block PodBrowser-icon PodBrowser-iconBlocked"
      />
      <div
        class="PodBrowser-textContainer"
      >
        <div
          class="PodBrowser-titleAndButtonContainer"
        >
          <p
            class="PodBrowser-title"
          >
            Blocked
          </p>
          <button
            class="PodBrowser-button PodBrowser-button--action PodBrowser-button"
            data-testid="add-agent-button"
          >
            <i
              alt="Block Button"
              class="PodBrowser-icon-add PodBrowser-icon"
            />
            Block
          </button>
        </div>
        <span
          class="PodBrowser-description"
        >
          <p>
            <b>
              Cannot 
            </b>
            view this resource
          </p>
        </span>
      </div>
    </div>
    <div
      class="PodBrowser-container PodBrowser-entered"
      style="min-height: 0px;"
    >
      <div
        class="PodBrowser-wrapper"
      >
        <div
          class="PodBrowser-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="PodBrowser-permissionsContainer"
            >
              <span
                class="PodBrowser-emptyStateTextContainer"
              >
                <p>
                  No one is blocked
                </p>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
